src/java.base/share/classes/java/math/BigInteger.java

```diff
@@ -2701,6 +2694,130 @@ public BigInteger pow(int exponent) {
         }
     }
 
+    /**
+     * Computes {@code x^n} using repeated squaring trick.
+     * Assumes {@code x != 0}.
+     */
+    private static BigInteger unsignedIntPow(int x, int n) {
+        if (x == 1)
+            return ONE;
+        // Double.PRECISION / bitLength(x) is the largest integer e
+        // such that x^e fits into a double. If e <= 2, we won't use fp arithmetic.
+        // This allows to use fp arithmetic where possible.
+        final int maxExp = Math.max(2, Double.PRECISION / bitLengthForInt(x));
+        final int maxExpLen = bitLengthForInt(maxExp);
+        final BigInteger[] powerCache = new BigInteger[1 << maxExpLen];
+
+        final int nZeros = Integer.numberOfLeadingZeros(n);
+        n <<= nZeros;
+
+        BigInteger pow = ONE;
+        int blockLen;
+        for (int nLen = Integer.SIZE - nZeros; nLen > 0; nLen -= blockLen) {
+            blockLen = maxExpLen < nLen ? maxExpLen : nLen;
+            // compute pow^(2^blockLen)
+            if (!pow.equals(ONE)) {
+                for (int i = 0; i < blockLen; i++)
+                    pow = pow.multiply(pow);
+            }
+
+            // add exp to power's exponent
+            int exp = n >>> -blockLen;
+            if (exp > 0) {
+                BigInteger xToExp = powerCache[exp];
+                if (xToExp == null) {
```

Comment:
Can we extract the cache computation logic into a new method for clarity? Huge methods hamper readability.

This use site will look like:
```java
if (exp > 0) {
    BigInteger xToExp = powerCache[exp];
    if (xToExp == null) {
        xToExp = computePower(powerCache, x, exp, maxExp);
    }
}
```